chainerx_cc/chainerx/native/native_device/reduction.cc

```diff
@@ -106,6 +106,73 @@ class NativeSumKernel : public SumKernel {
 
 CHAINERX_NATIVE_REGISTER_KERNEL(SumKernel, NativeSumKernel);
 
+class NativeCumsumKernel : public CumsumKernel {
+public:
+    void Call(const Array& a, int8_t axis, const Array& out) override {
+        a.device().CheckDevicesCompatible(a, out);
+
+        VisitDtype(out.dtype(), [&a, axis, &out](auto pt) {
+            using T = typename decltype(pt)::type;
+
+            IndexableArray<const T> a_iarray{a};
+            IndexableArray<T> out_iarray{out};
+            Indexer<> out_indexer{out.shape()};
+            Indexer<> prev_indexer{a.shape()};
+
+            int64_t axis_dim = a.shape()[axis];
+
+            // left: set of input dimensions lower than the axis
+            // right: set of input dimensions higher than the axis
+            Shape left_shape{a.shape().begin(), a.shape().begin() + axis};
+            Shape right_shape{a.shape().begin() + (axis + 1), a.shape().end()};
+            Shape axis_shape{axis_dim};  // always ndim==1
+
+            Indexer<> left_indexer{left_shape};
+            Indexer<> right_indexer{right_shape};
+            Indexer<> axis_indexer{axis_shape};
+            Indexer<> indices_indexer{axis_shape};
+
+            auto it_left = left_indexer.It(0);
+            auto it_right = right_indexer.It(0);
+            auto it_axis = axis_indexer.It(0);
+            auto it_out = out_indexer.It(0);
+            auto it_prev = prev_indexer.It(0);
+
+            // Copy
+            for (auto it = out_indexer.It(0); it; ++it) {
+                out_iarray[it] = a_iarray[it];
+            }
+
+            for (auto it = indices_indexer.It(1); it; ++it) {
+                int64_t index = it.raw_index();
+                index = index % axis_dim;
+                CHAINERX_ASSERT(0 <= index);
+                CHAINERX_ASSERT(index < axis_dim);
+
+                it_axis.Restart(index);
+
+                it_out.CopyIndex(it, it_left.ndim());
+                it_prev.CopyIndex(it_axis, it_left.ndim());
+
+                for (it_left.Restart(); it_left; ++it_left) {
+                    it_out.CopyIndex(it_left);
+                    it_prev.CopyIndex(it_left);
+
+                    for (it_right.Restart(); it_right; ++it_right) {
+                        it_out.CopyIndex(it_right, it_left.ndim() + it.ndim());
+                        it_prev.CopyIndex(it_right, it_left.ndim() + it_axis.ndim());
+                        it_prev.index()[axis] -= 1;
+                        out_iarray[it_out] += out_iarray[it_prev];
```

Comment:
float16 arrays store their items as `uint16_t` values because C++ and CUDA don't have the float16 type as a primitive type.
https://github.com/chainer/chainer/blob/c47a3785bca0e36af35406f43776004ad1030da1/chainerx_cc/chainerx/dtype.h#L108
So, before calculating summation, we need to convert types with `native_internal::StorageToDataType` defined in `chainerx/native/data_type.h`.
```suggestion
                        native_internal::StorageToDataType<T>(out_iarray[it_out]) +=
                                native_internal::StorageToDataType<T>(out_iarray[it_prev]);

```

Reply:
Thanks for the explanation.

Reply:
Please include `data_type.h`

Reply:
Done.

Reply:
Done.